### PR TITLE
Add gradient clipping

### PR DESCRIPTION
**Pull request by srai9**
I added an option to enable gradient clipping. It made training more stable in my experiments, especially when using smaller batch sizes.

You can see the effect in these logs (purple = with gradient clipping, green = without).
grad_clip

It's disabled by default and can be enabled by setting `max_grad_norm` in `config.py`

### CODE REVIEW

**Comment by lusxvr** [comment]
image
 When I plot my grad_norm (pre_clipping, but still gets clipped before applying the gradients), I still get spikes in the loss even after clipping it to 1.0. Do you know what could be the cause of this?

**Review by andimarafioti — COMMENTED** [review]
Great job! I wanted to do this! Can we also add the grad_norm to wandb?

**Comment by srai9** [comment]
Thanks for checking it out :) I've updated the code to add grad_norm to wandb.

**Review by lusxvr — APPROVED** [review]
LGTM! Love the handling of logging dependent on if we clip the gradients or not, very clean!
